lib/pacemaker/pcmk_verify.c

```diff
@@ -0,0 +1,172 @@
+
+int
+pcmk__verify(pcmk_scheduler_t *scheduler, pcmk__output_t *out,
+                   gboolean use_live_cib, const char *xml_file,
+                   const char *xml_string, gboolean xml_stdin,
+                   const char *cib_save, int verbosity, const char* output_ty)
+{
+    int rc = pcmk_rc_ok;
+    GError *error = NULL;
+    xmlNode *cib_object = NULL;
+    xmlNode *status = NULL;
+
+    crm_info("=#=#=#=#= Getting XML =#=#=#=#=");
+
+    if (use_live_cib) {
+        crm_info("Reading XML from: live cluster");
+        rc = cib__signon_query(out, NULL, &cib_object);
+
+        if (rc != pcmk_rc_ok) {
+            // cib__signon_query() outputs any relevant error
+        }
+
+    } else if (xml_file != NULL) {
+        cib_object = filename2xml(xml_file);
+        if (cib_object == NULL) {
+            rc = ENODATA;
+            g_set_error(&error, PCMK__RC_ERROR, rc, "Couldn't parse input file: %s", xml_file);
+        }
+
+    } else if (xml_string != NULL) {
+        cib_object = string2xml(xml_string);
+        if (cib_object == NULL) {
+            rc = ENODATA;
+            g_set_error(&error, PCMK__RC_ERROR, rc, "Couldn't parse input string: %s", xml_string);
+        }
+    } else if (xml_stdin) {
+        cib_object = stdin2xml();
+        if (cib_object == NULL) {
+            rc = ENODATA;
+            g_set_error(&error, PCMK__RC_ERROR, rc, "Couldn't parse input from STDIN.");
+        }
+
+    } else {
+        rc = ENODATA;
+        g_set_error(&error, PCMK__RC_ERROR, rc,
+                    "No configuration source specified.  Use --help for usage information.");
+    }
+
+    if (!pcmk__xe_is(cib_object, XML_TAG_CIB)) {
+        rc = EBADMSG;
+        g_set_error(&error, PCMK__RC_ERROR, rc,
+                    "This tool can only check complete configurations (i.e. those starting with <cib>).");
+    }
+
+    if (cib_save != NULL) {
+        write_xml_file(cib_object, cib_save, FALSE);
+    }
+
+    status = pcmk_find_cib_element(cib_object, XML_CIB_TAG_STATUS);
+    if (status == NULL) {
+        create_xml_node(cib_object, XML_CIB_TAG_STATUS);
+    }
+
+    if (pcmk__validate_xml(cib_object, NULL, (xmlRelaxNGValidityErrorFunc) out->err, out) == FALSE) {
+        pcmk__config_err("CIB did not pass schema validation");
+        free_xml(cib_object);
+        cib_object = NULL;
+
+    } else if (cli_config_update(&cib_object, NULL, FALSE) == FALSE) {
+        crm_config_error = TRUE;
+        free_xml(cib_object);
+        cib_object = NULL;
+        out->err(out, "The cluster will NOT be able to use this configuration.\n"
+                 "Please manually update the configuration to conform to the %s syntax.",
+                 xml_latest_schema());
+    }
+
+    scheduler = pe_new_working_set();
+    if (scheduler == NULL) {
+        rc = errno;
+        g_set_error(&error, PCMK__RC_ERROR, rc,
+                    "Could not allocate scheduler data: %s", pcmk_rc_str(rc));
+    }
+    scheduler->priv = out;
```

Comment:
I think throughout this function, you should keep the behavior from crm_verify.c where there's a `done` label that gets jumped to for all error cleanup and returning.  Otherwise, you've got to remember to protect everything that happens later in the function.  For instance, here you're going to have a segfault if `pe_new_working_set` fails because you'll dereference a NULL.

In general, I find jumping to a single point at the end of the function much easier to read than having to guard everything, even if we were all trained that goto was terrible.